### ISSUE TITLE
Video download slow since 19th May

### ISSUE THREAD

**Issue by LaticePrime** [instaloader/instaloader#1145]
Has anyone else found downloading profile posts unusably slow for the last two days? Scraping stories still just fine!

**Comment by skrull** [comment]
Yes, same here.

**Comment by LaticePrime** [comment]
Thanks! Good to know it's not just me!

**Comment by aandergr** [comment]
Interestingly, just 4 days before this issue we received a bug report ["Instaloader works too fast"](https://github.com/instaloader/instaloader/issues/1141). 

Since July 2020, Instaloader's rate controller can be easily customized without having to touch Instaloader's code. So luckily, if you find Instaloader downloading too slow, too fast or that its perceived download speed varies too much, you can easily adapt it to match your expectations: https://instaloader.github.io/module/instaloadercontext.html#ratecontroller

**Comment by ekalin** [comment]
Same here.

It does not seem to have anything to do with the rate controller. The actual download of media is really slow. I can see the partially downloaded '.temp' files growing very slowly. It seems Instagram has added to throttle on their side.

**Comment by skrull** [comment]
It's something server side. I'm using 4.7.1 and this behavior started a few days ago with video files. It takes around 4 minutes to download each video here.

**Comment by LaticePrime** [comment]
This is exactly what's happened to me; when I view the .temp file it's now growing much more slowly than normal for profile videos only; snatching stories is still really fast. It was fine on Wednesday the 19th so started on the 20th. Has to be instagram side as I've changed nothing and running latest version.

**Comment by Lex-2008** [comment]
Just started using instaloader today, to learn that I was 4 days late :) Have anyone found a workaround? Does being logged in or using a different user-agent help?

**Comment by Lex-2008** [comment]
Note that with decreased speed instaloader _might_ fail to download a complete video file, leaving you with a partially-downloaded one!

It looks like issue is indeed in user-agent string. Probably the one used in instaloader got banned by Instagram. A working **WORKAROUND** is this:
1. visit echo.opera.com with a browser to see your user-agent
2. add `--user-agent` argument to your instaloader command

Alternatively, you can use my user-agent (Microsoft Edge Beta on Linux):
```
--user-agent 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.57 Safari/537.36 Edg/91.0.864.27'
```

**Comment by skrull** [comment]
Uh, much easier than I thought. Thanks.

**Comment by LaticePrime** [comment]
Didn't work for me 😞

**Comment by cschwem2er** [comment]
> Interestingly, just 4 days before this issue we received a bug report ["Instaloader works too fast"](https://github.com/instaloader/instaloader/issues/1141).
> 
> Since July 2020, Instaloader's rate controller can be easily customized without having to touch Instaloader's code. So luckily, if you find Instaloader downloading too slow, too fast or that its perceived download speed varies too much, you can easily adapt it to match your expectations: https://instaloader.github.io/module/instaloadercontext.html#ratecontroller

Many thanks for the clarification. What I didn't quite get is how to properly implement it though.

```import instaloader

class MyRateController(instaloader.RateController):
    def sleep(self, 5):
        raise MyCustomException()

L = instaloader.Instaloader(rate_controller=lambda ctx: MyRateController(ctx))
```
In this example, would instaloader wait for 5 seconds between each single call? And what is the default waiting time of instaloader in seconds? I would like to avoid running into the "too many calls" exception which I am getting quite frequently.

**Comment by aandergr** [comment]
Thanks to you all for sharing your observations! In my experiments, the download speed was fine and seemed not to depend on the user agent, however, sometimes these things might depend on the geographical region or other factors.

Does this problem still occur, or does changing the user agent still lead to faster downloads?

For reliably comparing download times, there is a flag `--no-sleep`, which removes all random delays between requests, while still maintaining the delays that the rate controller considers necessary. 


> Many thanks for the clarification. What I didn't quite get is how to properly implement it though.
> 
> ```
> 
> class MyRateController(instaloader.RateController):
>     def sleep(self, 5):
>         raise MyCustomException()
> 
> L = instaloader.Instaloader(rate_controller=lambda ctx: MyRateController(ctx))
> ```
> 
> In this example, would instaloader wait for 5 seconds between each single call? And what is the default waiting time of instaloader in seconds? I would like to avoid running into the "too many calls" exception which I am getting quite frequently.

Instagram has a set of rules regarding how many requests of a certain type are permitted in a certain time frame. We regularly do a bunch of incredibly time-intensive experiments to confirm or update our belief of how their rate limits work. The rate controller keeps track of the requests that Instaloader makes and implements delays to remain within Instagram's limits while still providing a reasonably fast experience.

If you want to add an *additional* delay of 5 seconds before every request, the following snippet could be used

```python
import instaloader

class MyRateController(instaloader.RateController):
    def query_waittime(self, query_type, current_time, untracked_queries=False):
        return 5 + super().query_waittime(query_type, current_time, untracked_queries)

L = instaloader.Instaloader(rate_controller=lambda ctx: MyRateController(ctx))
```

**Comment by ekalin** [comment]
The slow download is still affecting me. I've tried to change my user agent to the same as my browser (Firefox), but it didn't work.

I didn't try with `--no-sleep`, but as I mentioned before, it's the download of media that's slow. Not so noticeable with pictures because they're small anyway, but very noticeable with videos, especially larger ones. I can tell that the slowness is in the download itself because I can see the `.temp` file, and its very slow growth.

**Comment by ItsABlackScreen** [comment]
The slow profile download is a problem for me too. 
Things I have tried: 
1. Changed the --user-agent string to my browser(firefox), that didn't change anything.
2. Tried with --no-sleep but no change.
3. Changed accounts, with no change.

The issue is the same as others with very slow .temp file growth so, it's a download speed issue. Stories download works with fast speeds as before and in regards to being rate limited, the profile i usually use for scraping was last used almost a month ago with Instaloader.

**Comment by LaticePrime** [comment]
Unfortunately download of videos from profiles still very very slow rendering instaloader pretty much unusable until there is a fix 😞  

All stories and pictures fine.

Have tried everything mentioned here and more and nothing has worked.

Started on the 20th of May.

**Comment by skrull** [comment]
I have one account that doesn't exhibit this behavior, and I couldn't figure (yet) what triggers video slowdown. Maybe accounts are getting flagged for this permanently?

**Comment by LaticePrime** [comment]
I'm not logged in when mine does it, but logging in makes no difference. Maybe IP is flagged. Although videos load at normal speed in app and on browser.

**Comment by aandergr** [comment]
Then let's track this as a *bug* rather than a question.

Also related to video downloading is #1022 (there are multiple video versions with different URLs available), however all video versions available in the structures we fetch seem to be served very slowly.

**Comment by Neurotoxin001** [comment]
Maybe it can be fixed by using iOS or Android IG app useragent or instagram.com referrer?

**Comment by Neurotoxin001** [comment]
No news about fix it?

**Comment by gavvvr** [comment]
If you need an urgent solution:
- The following project got the same problem fixed: https://github.com/HuiiBuh/InstagramDownloader/pull/205
- I also noticed the following project did not have such a problem: https://github.com/adw0rd/instagrapi (probably because it uses private API)

**Comment by skrull** [comment]
Changing the user-agent string solves this for me. Maybe your account/ip got flagged and is slow on purpose?

**Comment by LaticePrime** [comment]
After reading #1233 I tried L.context.user_agent = None and this fixed the speed problem for me

**Comment by LaticePrime** [comment]
I also just placed # in front of Line 155 del header['Referer'] in instaloadercontext.py and this fixed the speed problem as well

**Comment by aandergr** [comment]
This is fixed in [Version 4.7.5](https://github.com/instaloader/instaloader/releases/tag/v4.7.5). Thanks to @fireattack for discovering and contributing the solution!

**Comment by estatistics** [comment]
how you insert this snipset and where? in new file? or inside instaloader.py? 
How you invoke that behavior? 
Cause using `python3  instaloader_wait.py` get relative path and circular errors.
